### ISSUE TITLE
Returning a different error when client receives CloseNormalClosure

### ISSUE THREAD

**Issue by dsabeti** [gorilla/websocket#70]
Hi,

We're trying to have our websocket server send informational error messages when it sends a Close control frame (as specified [here](https://tools.ietf.org/html/rfc6455#section-5.5.1)). Gorilla even supports this in the `WriteControl` function.

But in the `ReadMessage` function, the Close frame's message body gets clobbered by an `io.EOF` error (I guess, strictly speaking, this happens in `advanceFrame` but the public method that gets the bad behavior is `ReadMessage`). You can see this code here: https://github.com/gorilla/websocket/blob/master/conn.go#L676-L680

We'd like our websocket client to be able to display the Close frame's message body at least for the type `CloseNormalClosure`, but we're not sure all the implications of changing this behavior. Maybe the following implementation would work:
```
		switch closeCode {
		case CloseNormalClosure:
			return noFrame, &closeError{code: closeCode, text: closeText}
                case CloseGoingAway:
                        return noFrame, io.EOF
		default:
			return noFrame, &closeError{code: closeCode, text: closeText}
		}
```

We're happy to submit a PR if we get a go-ahead from someone, but we figured we'd start with a discussion.

Thanks,
David && Warren

**Comment by garyburd** [comment]
The goal of the current code is to convert the normal closes cases into a single error that the application can ignore.  The application might want to log all other errors.

I am not in favor of the change proposed in the issue. First, it does not provide an easy way for applications to distinguish normal closure from other closure. Second, If any errors are converted to io.EOF, Normal Closure should be included. It's the error closest in meaning to io.EOF.

An alternative approach is to return closeError for all close frames and add the function `IsCleanClose(err error) bool`. With this feature, the application can write:

    if !websocket.IsCleanClose(err) {
        // log the error, increment error counter, ...
   }

My proposed change and the change proposed in the issue might break existing applications. I'll survey existing uses to see how apps are handling errors.

**Comment by dsabeti** [comment]
Fair enough. All we need is to be able to get the message body associated with the close frame. Just to clarify, does this implementation mean that gorilla never returns EOF for close frames?

**Comment by garyburd** [comment]
With my proposed change, io.EOF will not be returned for close frames. I am uncertain if my proposed change is a good idea. I found applications that check for io.EOF.

Yet another approach is to return a closeError when reason != "".

If I do go through with the proposed change, I'll need to find a better name for IsCleanClose.  The browser client API has a `wasClean` field with a different meaning.

**Comment by tsavola** [comment]
I would be interested in the close code (number) of a closeError.  Currently even the error type is not public.

**Comment by garyburd** [comment]
Fixed by b6ab76f1fe9803ee1d59e7e5b2a797c1fe897ce5
